### PR TITLE
Add support of typescript for panel plugin

### PR DESCRIPTION
**Pull request by amper43**
* typescript support for Panel plugin added

### CODE REVIEW

File: src/project_schema/index.ts [hunk 1]
```diff
@@ -2,12 +2,20 @@ import build from './build';
...
 export default new FolderGenerator<TemplateOptions>(context => context.options.id, [
-  new TemplateGenerator(require('./package.json.ejs'), 'package.json'),
+  new TemplateGenerator(require('./package.json.ejs'),
+                        'package.json',
+                        (ctx) => { ctx['extension'] = srcExt('{ext}')(ctx) }
+  ),
   new TemplateGenerator(require('./.gitignore.ejs'), '.gitignore'),
   new TemplateGenerator(require('./README.md.ejs'), 'README.md'),
+  context => {
+    if(context.options.pluginType === PluginType.Panel && context.options.language === Language.TypeScript) {
```
Comment (rozetko) [inline src/project_schema/index.ts]: `tsconfig` for panel is the same as for datasource
So I think we can check only for language

Comment (amper43) [reply src/project_schema/index.ts]: @rozetko  but if only JavaScript required?

File: src/project_schema/build/index.ts [hunk 2]
```diff
@@ -1,8 +1,12 @@
-import { FolderGenerator, TemplateGenerator } from 'src/generators';
+import { FolderGenerator, TemplateGenerator, GenerationContext } from 'src/generators';
+import { srcExt, TemplateOptions } from 'src/template_options';
 
 
 export default new FolderGenerator('build', [
-  new TemplateGenerator(require('./webpack.base.conf.js.ejs'), 'webpack.base.conf.js'),
+  new TemplateGenerator(require('./webpack.base.conf.js.ejs'),
+  'webpack.base.conf.js',
```
Comment (jonyrock) [inline src/project_schema/build/index.ts]: indent

File: src/project_schema/build/index.ts [hunk 3]
```diff
@@ -1,8 +1,12 @@
-import { FolderGenerator, TemplateGenerator } from 'src/generators';
+import { FolderGenerator, TemplateGenerator, GenerationContext } from 'src/generators';
+import { srcExt, TemplateOptions } from 'src/template_options';
 
 
 export default new FolderGenerator('build', [
-  new TemplateGenerator(require('./webpack.base.conf.js.ejs'), 'webpack.base.conf.js'),
+  new TemplateGenerator(require('./webpack.base.conf.js.ejs'),
+  'webpack.base.conf.js',
+  (ctx: GenerationContext<TemplateOptions>) => { ctx['extension'] = srcExt('{ext}')(ctx) }
```
Comment (jonyrock) [inline src/project_schema/build/index.ts]: - indent
```suggestion
  ctx => { ctx['extension'] = srcExt('{ext}')(ctx) }
```

File: src/project_schema/index.ts [hunk 4]
```diff
@@ -2,12 +2,20 @@ import build from './build';
 import src from './src';
 
 import { FolderGenerator, TemplateGenerator } from 'src/generators';
-import { TemplateOptions } from 'src/template_options';
+import { TemplateOptions, PluginType, Language, srcExt } from 'src/template_options';
 
 
 export default new FolderGenerator<TemplateOptions>(context => context.options.id, [
-  new TemplateGenerator(require('./package.json.ejs'), 'package.json'),
+  new TemplateGenerator(require('./package.json.ejs'),
```
Comment (jonyrock) [inline src/project_schema/index.ts]: bad indention , very bad

File: src/project_schema/package.json.ejs [hunk 5]
```diff
@@ -12,6 +12,11 @@
   "license": "MIT",
   "repository": "",
   "devDependencies": {
+    <% if (extension === 'ts') { %>
```
Comment (jonyrock) [inline src/project_schema/package.json.ejs]: no no, please set `context['isTypeScript']` in contextModifier

File: src/project_schema/build/index.ts (hunk 3 above)
Comment (amper43) [reply src/project_schema/build/index.ts]: it throw error:
```
[ts]
Argument of type 'GenerationContext<{}>' is not assignable to parameter of type 'GenerationContext<TemplateOptions>'.
  Type '{}' is not assignable to type 'TemplateOptions'.
    Property 'id' is missing in type '{}'.
```

Comment (jonyrock) [reply src/project_schema/build/index.ts]: you should write `TemplateGenerator<TemplateOptions>` then

**Pull request merged**